### PR TITLE
Add color mode to ui store

### LINKED ISSUE #4308: Add frontend "color mode" state to the ui cookie
> This issue is part of the Dark Mode project: #3592. Please see the [implementation plan](https://docs.openverse.org/projects/proposals/dark_mode/20240325-implementation_plan_dark_mode.html) for additional detail and context. 

> [!IMPORTANT]
> The implementation plan originally specified storing a `systemValue` representing the actual media query value of "prefers-color-scheme" in the color mode state. This was deemed unnecessary and removed in this issue.
>
> The implementation plan also called for, eventually, adding detection in our infrastructure to bypass caching when a dark/light mode _cookie_ was present for the user. By using the existing ui cookie, we bypass the need for any infrastructure changes here.

Update the frontend UI store to include a colorMode. This is a representation of whether the site should display in "light", "dark", or "system" (the default) color mode. We need functionality to:

- Retrieve the current color mode from the ui cookie. Override this preference if the `force_dark_mode` feature flag from #4303 is on.
- Set the color mode, saving it to the ui cookie. The eventual dark mode UI toggle will use this method.
- Additionally, the mechanism to set the body class `.light-mode` or `.dark-mode` added in #4303 needs to be updated to read from the ui store.

Here is a diff that should represent the necessary changes to the store and cookie. Please update tests as well to accompany these changes.

```diff
diff --git a/frontend/src/stores/ui.ts b/frontend/src/stores/ui.ts
index 6c09abb64..650f5548a 100644
--- a/frontend/src/stores/ui.ts
+++ b/frontend/src/stores/ui.ts
@@ -8,10 +8,14 @@ import { ALL_SCREEN_SIZES } from "~/constants/screens"
 import { cookieOptions } from "~/utils/cookies"
 import { needsTranslationBanner } from "~/utils/translation-banner"
 
+import { useFeatureFlagStore } from "./feature-flag"
+
 import type { LocaleObject } from "@nuxtjs/i18n"
 
 const desktopBreakpoints: RealBreakpoint[] = ["2xl", "xl", "lg"]
 
+type ColorMode = "dark" | "light" | "system"
+
 export interface UiState {
   /**
    * whether to show the instructions snackbar.
@@ -47,6 +51,8 @@ export interface UiState {
   shouldBlurSensitive: boolean
   /* A list of sensitive single result UUIDs the user has opted-into seeing */
   revealedSensitiveResults: string[]
+  /* Whether the UI is using dark, light, or system mode */
+  colorMode: ColorMode
 }
 
 export const breakpoints = Object.keys(ALL_SCREEN_SIZES)
@@ -61,6 +67,7 @@ export const useUiStore = defineStore("ui", {
     dismissedBanners: [],
     shouldBlurSensitive: true,
     revealedSensitiveResults: [],
+    colorMode: "system",
   }),
 
   getters: {
@@ -70,6 +77,7 @@ export const useUiStore = defineStore("ui", {
         isFilterDismissed: state.isFilterDismissed,
         breakpoint: state.breakpoint,
         dismissedBanners: Array.from(this.dismissedBanners),
+        colorMode: state.colorMode,
       }
     },
     areInstructionsVisible(state): boolean {
@@ -125,6 +133,14 @@ export const useUiStore = defineStore("ui", {
     shouldShowAnalyticsBanner(): boolean {
       return !this.dismissedBanners.includes("analytics")
     },
+    currentColorMode(state): ColorMode {
+      const featureFlagStore = useFeatureFlagStore()
+      if (featureFlagStore.isOn("force_dark_mode")) {
+        return "dark"
+      }
+      // Return the cookie value otherwise
+      return state.colorMode
+    },
   },
 
   actions: {
@@ -176,6 +192,13 @@ export const useUiStore = defineStore("ui", {
         this.dismissedBanners = cookies.dismissedBanners
       }
 
+      if (
+        typeof cookies.colorMode === "string" &&
+        ["dark", "light", "system"].includes(cookies.colorMode)
+      ) {
+        this.colorMode = cookies.colorMode as ColorMode
+      }
+
       this.writeToCookie()
     },
     /**
@@ -255,5 +278,9 @@ export const useUiStore = defineStore("ui", {
       this.shouldBlurSensitive = value
       this.revealedSensitiveResults = []
     },
+    setColorMode(preference: "dark" | "light" | "system") {
+      this.colorMode = preference
+      this.writeToCookie()
+    },
   },
 })
diff --git a/frontend/src/types/cookies.ts b/frontend/src/types/cookies.ts
index b64d419ca..6acd86e72 100644
--- a/frontend/src/types/cookies.ts
+++ b/frontend/src/types/cookies.ts
@@ -39,6 +39,10 @@ export interface OpenverseCookieState {
      * The list of ids of dismissed banners.
      */
     dismissedBanners?: BannerId[]
+    /**
+     * A string representing the color scheme the site should use.
+     */
+    colorMode?: string
   }
   /**
    * The state of the persistent feature flags.

```

### PR DESCRIPTION
**Pull request by zackkrida**
## Fixes

Fixes #4308 by @zackkrida 

## Description

Core functionality:

- Updated the `dark_mode_ui_toggle` feature flag description to reflect that it now causes the frontend to respect system preferences for color themes. Now when the color mode is "system", no body css class will be set.
- Added media queries in the tailwind.css to adjust the color scheme based on the system's prefers-color-scheme setting.
- Introduced a new `colorMode` state in the ui store which stores the color mode in a cookie.
- Added an action to set the color mode, which the toggle in #4309 will use.
- Updated the `useDarkMode` composable to derive the color mode from the UI store when the `dark_mode_ui_toggle` is enabled, and to override it with the `force_dark_mode` flag when enabled.
- Updated and expanded unit tests for the `useDarkMode` composable and the ui store to cover the new `colorMode` logic.

I also made some small improvements:

- Refactored the UI store's initial state setup to use a new defaultUiState object, which can be used to initialize tests as well.
- Updated type definitions in the feature flag store for stricter type checking (FlagName instead of string for various methods).
- Converted the ui store test to TypeScript and fixed some types.

Here is a video outlining the feature flag behavior:

## Testing Instructions

- Review the code changes and test cases to make sure they are sufficient
- Enable the preference to force dark mode; observe that it is still respected
- Apply this diff to add a rudimentary theme changer to the footer:
```diff
diff --git a/frontend/src/components/VFooter/VFooter.vue b/frontend/src/components/VFooter/VFooter.vue
index ca326f8f4..24d699757 100644
--- a/frontend/src/components/VFooter/VFooter.vue
+++ b/frontend/src/components/VFooter/VFooter.vue
@@ -14,6 +14,7 @@ import VBrand from "~/components/VBrand/VBrand.vue"
 import VLanguageSelect from "~/components/VLanguageSelect/VLanguageSelect.vue"
 import VPageLinks from "~/components/VHeader/VPageLinks.vue"
 import VWordPressLink from "~/components/VHeader/VWordPressLink.vue"
+import { useFeatureFlagStore } from "#imports"
 
 /**
  * The footer is the section displayed at the bottom of a page. It can contain
@@ -72,11 +73,24 @@ export default defineComponent({
       "--link-col-height": Math.ceil(Object.keys(allPages).length / 2),
     }))
 
+    const featureFlagStore = useFeatureFlagStore()
+    const showToggle = computed(() =>
+      featureFlagStore.isOn("dark_mode_ui_toggle")
+    )
+    const colorMode = ref(uiStore.colorMode)
+    const handleColorModeChange = () => {
+      uiStore.setColorMode(colorMode.value)
+    }
+
     return {
       isContentMode,
       allPages,
       currentPage,
 
+      showToggle,
+      colorMode,
+      handleColorModeChange,
+
       footerEl,
       variantNames,
       linkColumnHeight,
@@ -110,6 +124,17 @@ export default defineComponent({
 
     <!-- Locale chooser and WordPress affiliation graphic -->
     <div class="locale-and-wp flex flex-col justify-between">
+      <select
+        class="bg-tx"
+        v-if="showToggle"
+        v-model="colorMode"
+        @change="handleColorModeChange"
+        name="colorMode"
+      >
+        <option value="light">Light</option>
+        <option value="dark">Dark</option>
+        <option value="system">System</option>
+      </select>
       <VLanguageSelect
         v-bind="languageProps"
         class="language max-w-full border-secondary"
```
- Enable the `dark_mode_ui_toggle` flag on the preferences page.
- Try each setting in the footer theme toggle and make sure:
    - Choice is persisted with the correct CSS class on "body" and the correct value in the Ui cookie (check devtools for this)
    - The "system" setting respects if you toggle dark mode at your OS or browser level 

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [x] I ran the DAG documentation generator (`ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`ov just catalog/generate-docs media-props`
      for the catalog or `ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: frontend/src/stores/ui.ts [hunk 1]
```diff
@@ -257,6 +281,10 @@ export const useUiStore = defineStore("ui", {
       this.shouldBlurSensitive = value
       this.revealedSensitiveResults = []
     },
+    setColorMode(preference: "dark" | "light" | "system") {
```
Comment (obulat) [inline frontend/src/stores/ui.ts:284]:
```suggestion
    setColorMode(preference: ColorMode) {
```

**Review by obulat — COMMENTED** [review]
Just noting here that I don't remember when we save the cookies: if any preferences have been changed (on preferences page or using something like the sensitivity toggle) or when the specific feature was changed (so, the color mode cookie is saved only if you adjust the color mode cookie) .

File: frontend/src/stores/ui.ts [hunk 2]
```diff
@@ -17,6 +17,15 @@ import { needsTranslationBanner } from "~/utils/translation-banner"
...
 
+export type SnackbarState = "not_shown" | "visible" | "dismissed"
+export type ColorMode = "dark" | "light" | "system"
+
+export function isColorMode(value: undefined | string): value is ColorMode {
+  return (
+    typeof value === "string" && ["light", "dark", "system"].includes(value)
+  )
+}
+
```
Comment (zackkrida) [inline frontend/src/stores/ui.ts:28]: I moved these types in here; it makes more sense to me that the UI Store is the "source of truth" for UI state, and cookies are an implementation of a sub-set of that state.

File: frontend/src/stores/ui.ts [hunk 3]
```diff
@@ -53,22 +62,28 @@ export interface UiState {
   /* A list of sensitive single result UUIDs the user has opted-into seeing */
   revealedSensitiveResults: string[]
   headerHeight: number
+
+  /* Whether the UI is using dark, light, or system mode */
```
Comment (zackkrida) [inline frontend/src/stores/ui.ts:66]:
```suggestion
  /* The user-chosen color theme of the site. */
```

File: frontend/src/stores/ui.ts [hunk 4]
```diff
@@ -53,22 +62,28 @@ export interface UiState {
...
-    innerFilterVisible: false,
-    isFilterDismissed: false,
-    isDesktopLayout: false,
-    breakpoint: "sm",
-    dismissedBanners: [],
-    shouldBlurSensitive: true,
-    revealedSensitiveResults: [],
-    headerHeight: 80,
-  }),
+  state: (): UiState => ({ ...defaultUiState }),
```
Comment (zackkrida) [inline frontend/src/stores/ui.ts:86]: The spread here isn't _super_ necessary, but fixes an issue with tests wherin the imported defaultUiState was being mutated between tests. 

Logically this makes sense anyway, that the store would "copy in" all the properties of the default state object

File: frontend/src/stores/ui.ts [hunk 5]
```diff
@@ -182,7 +202,14 @@ export const useUiStore = defineStore("ui", {
         "ui",
         persistentCookieOptions
       )
-      uiCookie.value = this.cookieState
+
+      console.log(defaultPersistientCookieState.ui)
+      console.log(this.cookieState)
```
Comment (zackkrida) [inline frontend/src/stores/ui.ts:207]:
```suggestion
```

File: frontend/src/types/cookies.ts [hunk 6]
```diff
@@ -15,6 +12,12 @@ export const persistentCookieOptions = {
   maxAge: 60 * 60 * 24 * 60, // 60 days; Makes the cookie persistent.
 } as const
 
+export const defaultPersistientCookieState: OpenverseCookieState = {
+  ui: {
+    colorMode: "system",
+  },
+}
```
Comment (zackkrida) [inline frontend/src/types/cookies.ts:19]: This is a new place to configure default cookie values. Open to alternatives.

**Review by obulat — APPROVED** [review]
Very nice updates to TS! I agree with moving the variables around to more logical locations.
The color mode works well. The only thing I would like (not a blocker) is to turn the "force_dark_mode" feature flag off when you select the "light theme" (in the footer). Otherwise, I had to remove the session cookie for the force_dark_mode to see the light theme

**Pull request merged**